### PR TITLE
add symbol highlighting

### LINKED ISSUE #118: Add symbol highlighting

### PR DESCRIPTION
**Pull request by stas-vilchik**
Fixes #118

### CODE REVIEW

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 1]
```diff
@@ -281,6 +284,17 @@ private void saveHighlights(SensorContext sensorContext, Highlight[] highlights,
     highlighting.save();
   }
 
+  private void saveUsages(SensorContext sensorContext, Symbol[] symbols, InputFile inputFile) {
```
Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: Code Smell Code Smell: Make "saveUsages" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=stas%2Fsymbol-highlighting&issues=AWCdvfnyqDrCo5ysf9-v&open=AWCdvfnyqDrCo5ysf9-v)

[ISSUE_KEY]: AWCdvfnyqDrCo5ysf9-v

File: sonarts-core/src/runner/symbolHighlighting.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+export default function getSymbolHighlighting(sourceFile: ts.SourceFile, program: ts.Program) {
+  const symbols: SymbolHighlighting[] = [];
+  const symbolTable = SymbolTableBuilder.build(sourceFile, program);
+  symbolTable.getSymbols().forEach(symbol => {
+    const allUsages = symbolTable.allUsages(symbol);
+    const declaration = allUsages.find(usage => Boolean(usage.flags & UsageFlag.DECLARATION));
+    if (declaration) {
+      const textRange = getTextRange(declaration.node);
+      const references = allUsages.filter(usage => !(usage.flags & UsageFlag.DECLARATION));
```
Comment (vilchik-elena) [inline sonarts-core/src/runner/symbolHighlighting.ts]: @stas-vilchik can't we just remove from `allUsages` the `declaration`?

File: sonarts-core/tests/seTestUtils.ts [hunk 3]
```diff
@@ -91,7 +91,7 @@ function executeFromSource(source: string) {
   return { result, program };
 }
 
-function parse(source: string) {
+export function parse(source: string) {
```
Comment (vilchik-elena) [inline sonarts-core/tests/seTestUtils.ts]: @stas-vilchik may be we should move this one to parser.ts?

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java (hunk 1 above)
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: @stas-vilchik rename to `saveSymbolHighlighting`?

File: sonarts-core/src/runner/symbolHighlighting.ts (hunk 2 above)
Comment (stas-vilchik) [reply sonarts-core/src/runner/symbolHighlighting.ts]: we could, however I'm not a big fan of
```typescript
allUsages.splice(allUsages.indexOf(declaration), 1);
```

I prefer `filter`ing

**Pull request merged**